concordium-consensus/src/Concordium/GlobalState/Persistent/TreeState.hs

```diff
@@ -378,20 +379,15 @@ loadSkovPersistentData rp _treeStateDirectory pbsc = do
     GenesisBlock gd' -> return gd'
     _ -> logExceptionAndThrowTS (DatabaseInvariantViolation "Block at height 0 is not a genesis block.")
 
-  -- Get the last finalized block.
-  (_lastFinalizationRecord, lfStoredBlock) <- liftIO (getLastBlock _db) >>= \case
-      Left s -> logExceptionAndThrowTS $ DatabaseInvariantViolation s
-      Right r -> return r
-  -- Check that the block state for the last finalized block is not corrupt. The check works because
-  -- the `BlobRef` to the block state is stored in LMDB only after the block state is stored in the
-  -- blob store. If the reference points to a blob that cannot be read, the blob store can be
-  -- assumed to be corrupted.
-  elfBlob <- liftIO . try $
-    readBlobBSFromHandle (bscBlobStore . PBS.pbscBlobStore $ pbsc) (sbState lfStoredBlock)
-  case elfBlob :: Either IOError BS.ByteString of
-    Left _ -> logExceptionAndThrowTS $
-      DatabaseInvariantViolation "Last finalized block cannot be read. Blockstate database recovery required."
-    Right _ -> do
+  -- Unroll the treestate if the last finalized blockstate is corrupted. If the last finalized
+  -- blockstate is not corrupted, the treestate is unchanged.
+  unrollTreeStateWhile _db isBlockStateCorrupted >>= \case
+    Left e -> logExceptionAndThrowTS . DatabaseInvariantViolation $
+              "The block state database is corrupt. Recovery attempt failed: " <> e
+    Right (_lastFinalizationRecord, lfStoredBlock) -> do
```

Comment:
I can see that now you are not logging anything at all if you delete stuff from the database.

That is not OK and it will make support harder. It should be very easy to return a list of block hashes and finalization records that were deleted, so you could just log those here. If the list was empty then nothing would be logged.